Fix tests and make more tests for the tags controller

All tests in `FoldersControllerTest` are failing with:
```
Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'com.guillermonegrete.gallery.tags.TagsRepository' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
```

Also add more tests for each endpoint of `TagsController`

--- a/src/main/kotlin/com/guillermonegrete/gallery/data/Folder.kt
+++ b/src/main/kotlin/com/guillermonegrete/gallery/data/Folder.kt
@@ -7,11 +7,6 @@ data class Folder(
         val id: Long,
 )
 
-data class GetFolderResponse(
-        val name: String,
-        val folders: List<Folder>
-)
-
 data class PagedFolderResponse(
         val name: String,
         val page: SimplePage<Folder>

--- a/src/main/kotlin/com/guillermonegrete/gallery/tags/data/TagEntity.kt
+++ b/src/main/kotlin/com/guillermonegrete/gallery/tags/data/TagEntity.kt
@@ -12,7 +12,7 @@ import javax.persistence.*
 @Entity
 open class TagEntity(
     @Column(nullable = false, unique = true)
-    open val name: String = "",
+    open val name: String,
     @Column(name = "creation_date", nullable = false)
     open val creationDate: Instant = Instant.now().truncatedTo(ChronoUnit.SECONDS), // by default the db saves in seconds, truncate to avoid having different milliseconds
     /**

--- a/src/test/kotlin/com/guillermonegrete/gallery/FoldersControllerTest.kt
+++ b/src/test/kotlin/com/guillermonegrete/gallery/FoldersControllerTest.kt
@@ -1,16 +1,17 @@
 package com.guillermonegrete.gallery
 
-import com.guillermonegrete.gallery.data.MediaFile
-import com.guillermonegrete.gallery.data.MediaFolder
-import com.guillermonegrete.gallery.data.PagedFolderResponse
-import com.guillermonegrete.gallery.data.SimplePage
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.guillermonegrete.gallery.data.*
 import com.guillermonegrete.gallery.data.files.FileMapper
+import com.guillermonegrete.gallery.data.files.dto.ImageFileDTO
 import com.guillermonegrete.gallery.repository.MediaFileRepository
 import com.guillermonegrete.gallery.repository.MediaFolderRepository
 import com.guillermonegrete.gallery.services.FolderFetchingService
 import com.ninjasquad.springmockk.MockkBean
 import io.mockk.every
 import org.hamcrest.collection.IsCollectionWithSize.hasSize
+import org.junit.jupiter.api.Assertions.assertEquals
 import org.junit.jupiter.api.Test
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.beans.factory.annotation.Value
@@ -33,7 +34,11 @@ import java.net.InetAddress
 @TestPropertySource(properties = [
     "base.path=dummy"
 ])
-class FoldersControllerTest(@Autowired val mockMvc: MockMvc) {
+class FoldersControllerTest(
+    @Autowired val mockMvc: MockMvc,
+    @Autowired val mapper: FileMapper,
+    @Autowired val objectMapper: ObjectMapper
+) {
 
     // For the bean in the application class, don't want to run it in these tests
     @MockkBean(relaxed = true)
@@ -53,16 +58,20 @@ class FoldersControllerTest(@Autowired val mockMvc: MockMvc) {
     fun `Folders endpoint returns paged response model`(){
         val pageable = PageRequest.of(0, 20)
         val content = List(21) { MediaFolder("name$it", listOf(MediaFile("image.jpg")), 100L + it) }
-        every { mediaFolderRepository.findAll(pageable) } returns PageImpl(content.subList(0, 20), pageable, content.size.toLong())
+        val subList = content.subList(0, 20)
+        every { mediaFolderRepository.findAll(pageable) } returns PageImpl(subList, pageable, content.size.toLong())
 
-        val expected = PagedFolderResponse(path, SimplePage(emptyList(), 0, 0))
-        mockMvc.perform(get("/folders")).andDo(print()).andExpect(status().isOk)
-            .andExpect(jsonPath("$.name").value(expected.name))
-            .andExpect(jsonPath("$.page.items", hasSize<Array<Any>>(20)))
-            .andExpect(jsonPath("$.page.items[0].name").value("name0"))
-            .andExpect(jsonPath("$.page.items[0].coverUrl").value("http://$ipAddress/images/name0/image.jpg"))
-            .andExpect(jsonPath("$.page.totalPages").value(2))
-            .andExpect(jsonPath("$.page.totalItems").value(21))
+        val expected = PagedFolderResponse(
+            path,
+            SimplePage(subList.map { Folder(it.name, "http://$ipAddress/images/${it.name}/image.jpg", 1, it.id) },
+            2, content.size)
+        )
+        val result = mockMvc.perform(get("/folders")).andDo(print())
+            .andExpect(status().isOk)
+            .andReturn()
+
+        val resultResponse = objectMapper.readValue(result.response.contentAsString, PagedFolderResponse::class.java)
+        assertEquals(expected, resultResponse)
     }
 
     @Test
@@ -81,23 +90,24 @@ class FoldersControllerTest(@Autowired val mockMvc: MockMvc) {
     @Test
     fun `Sub folder returns page of image files`(){
         val subFolder = "subFolder"
-        val content = List(21) { MediaFile("image$it", 100 + it, 100 + it) }
-
         val mediaFolder = MediaFolder(subFolder)
-        val pageable = PageRequest.of(0, 20)
 
         every { mediaFolderRepository.findByName(subFolder) } returns mediaFolder
 
-        every { mediaFileRepository.findAllByFolder(mediaFolder, pageable) } returns PageImpl(content.subList(0, 20), pageable, content.size.toLong())
+        val pageable = PageRequest.of(0, 20)
+        val content = List(21) { MediaFile("image$it", 100 + it, 100 + it, folder = mediaFolder) }
+        val subList = content.subList(0, 20)
 
+        every { mediaFileRepository.findAllByFolder(mediaFolder, pageable) } returns PageImpl(subList, pageable, content.size.toLong())
+
+        val expected = SimplePage(subList.map { mapper.toDtoWithHost(it, ipAddress) }, 2, content.size)
         // First check the items then total pages and total items
-        mockMvc.perform(get("/folders/$subFolder").param("size", "20").param("page", "0")).andDo(print()).andExpect(status().isOk)
-            .andExpect(jsonPath("$.items", hasSize<Array<Any>>(20)))
-            .andExpect(jsonPath("$.items[0].url").value("http://$ipAddress/images/$subFolder/image0"))
-            .andExpect(jsonPath("$.items[0].width").value(100))
-            .andExpect(jsonPath("$.items[0].height").value(100))
-            .andExpect(jsonPath("$.totalPages").value(2))
-            .andExpect(jsonPath("$.totalItems").value(21))
+        val result = mockMvc.perform(get("/folders/$subFolder").param("size", "20").param("page", "0")).andDo(print())
+            .andExpect(status().isOk)
+            .andReturn()
+
+        val resultResponse = objectMapper.readValue(result.response.contentAsString, object: TypeReference<SimplePage<ImageFileDTO>>() {})
+        assertEquals(expected, resultResponse)
     }
 
     @TestConfiguration

--- a/src/test/kotlin/com/guillermonegrete/gallery/FoldersControllerTest.kt
+++ b/src/test/kotlin/com/guillermonegrete/gallery/FoldersControllerTest.kt
@@ -1,15 +1,15 @@
 package com.guillermonegrete.gallery
 
-import com.guillermonegrete.gallery.data.GetFolderResponse
 import com.guillermonegrete.gallery.data.MediaFile
 import com.guillermonegrete.gallery.data.MediaFolder
+import com.guillermonegrete.gallery.data.PagedFolderResponse
+import com.guillermonegrete.gallery.data.SimplePage
 import com.guillermonegrete.gallery.data.files.FileMapper
 import com.guillermonegrete.gallery.repository.MediaFileRepository
 import com.guillermonegrete.gallery.repository.MediaFolderRepository
 import com.guillermonegrete.gallery.services.FolderFetchingService
 import com.ninjasquad.springmockk.MockkBean
 import io.mockk.every
-import org.hamcrest.CoreMatchers.`is`
 import org.hamcrest.collection.IsCollectionWithSize.hasSize
 import org.junit.jupiter.api.Test
 import org.springframework.beans.factory.annotation.Autowired
@@ -24,10 +24,12 @@ import org.springframework.test.context.TestPropertySource
 import org.springframework.test.web.servlet.MockMvc
 import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
 import org.springframework.test.web.servlet.result.MockMvcResultHandlers.print
-import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
 import java.net.InetAddress
 
-@WebMvcTest
+
+@WebMvcTest(controllers = [FoldersController::class])
 @TestPropertySource(properties = [
     "base.path=dummy"
 ])
@@ -38,28 +40,34 @@ class FoldersControllerTest(@Autowired val mockMvc: MockMvc) {
     private lateinit var commandLineRunner: CommandLineRunner
     @MockkBean lateinit var service: FolderFetchingService
 
-    @MockkBean private lateinit var foldersRepository: FoldersRepository
     @MockkBean private lateinit var mediaFolderRepository: MediaFolderRepository
     @MockkBean private lateinit var mediaFileRepository: MediaFileRepository
 
     @Value("\${base.path}")
     private lateinit var path: String
 
+    private val ipAddress = InetAddress.getLocalHost().hostAddress
+
     @Throws(Exception::class)
     @Test
-    fun `Folders endpoint returns response model`(){
-        every { foldersRepository.getFolders(path) } returns emptyList()
+    fun `Folders endpoint returns paged response model`(){
+        val pageable = PageRequest.of(0, 20)
+        val content = List(21) { MediaFolder("name$it", listOf(MediaFile("image.jpg")), 100L + it) }
+        every { mediaFolderRepository.findAll(pageable) } returns PageImpl(content.subList(0, 20), pageable, content.size.toLong())
 
-        val expected = GetFolderResponse(path, emptyList())
+        val expected = PagedFolderResponse(path, SimplePage(emptyList(), 0, 0))
         mockMvc.perform(get("/folders")).andDo(print()).andExpect(status().isOk)
-                .andExpect(jsonPath("$.name").value(expected.name))
-                .andExpect(jsonPath("$.folders", `is`(expected.folders)))
+            .andExpect(jsonPath("$.name").value(expected.name))
+            .andExpect(jsonPath("$.page.items", hasSize<Array<Any>>(20)))
+            .andExpect(jsonPath("$.page.items[0].name").value("name0"))
+            .andExpect(jsonPath("$.page.items[0].coverUrl").value("http://$ipAddress/images/name0/image.jpg"))
+            .andExpect(jsonPath("$.page.totalPages").value(2))
+            .andExpect(jsonPath("$.page.totalItems").value(21))
     }
 
     @Test
     fun `Sub folder returns list of image files`(){
         val subFolder = "subFolder"
-        val ipAddress = InetAddress.getLocalHost().hostAddress
 
         every { mediaFolderRepository.findByName(subFolder) } returns MediaFolder(subFolder, listOf(MediaFile("image1", 100, 100)))
 
@@ -73,7 +81,6 @@ class FoldersControllerTest(@Autowired val mockMvc: MockMvc) {
     @Test
     fun `Sub folder returns page of image files`(){
         val subFolder = "subFolder"
-        val ipAddress = InetAddress.getLocalHost().hostAddress
         val content = List(21) { MediaFile("image$it", 100 + it, 100 + it) }
 
         val mediaFolder = MediaFolder(subFolder)

--- a/src/test/kotlin/com/guillermonegrete/gallery/tags/TagsControllerTest.kt
+++ b/src/test/kotlin/com/guillermonegrete/gallery/tags/TagsControllerTest.kt
@@ -1,23 +1,43 @@
 package com.guillermonegrete.gallery.tags
 
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.guillermonegrete.gallery.data.MediaFile
+import com.guillermonegrete.gallery.data.MediaFolder
+import com.guillermonegrete.gallery.data.SimplePage
 import com.guillermonegrete.gallery.data.files.FileMapper
+import com.guillermonegrete.gallery.data.files.ImageEntity
+import com.guillermonegrete.gallery.data.files.dto.ImageFileDTO
 import com.guillermonegrete.gallery.repository.MediaFileRepository
 import com.guillermonegrete.gallery.repository.MediaFolderRepository
 import com.guillermonegrete.gallery.tags.data.TagEntity
 import com.ninjasquad.springmockk.MockkBean
 import io.mockk.every
+import org.assertj.core.api.Assertions.assertThat
 import org.junit.jupiter.api.Test
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.boot.CommandLineRunner
 import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
 import org.springframework.boot.test.context.TestConfiguration
 import org.springframework.context.annotation.Bean
+import org.springframework.data.domain.PageImpl
+import org.springframework.data.domain.PageRequest
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.http.MediaType
 import org.springframework.test.web.servlet.MockMvc
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
 import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
+import org.springframework.test.web.servlet.result.MockMvcResultHandlers.print
 import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
+import java.net.InetAddress
+import java.util.*
 
 @WebMvcTest
-class TagsControllerTest(@Autowired val mockMvc: MockMvc) {
+class TagsControllerTest(
+    @Autowired val mockMvc: MockMvc,
+    @Autowired val mapper: FileMapper,
+    @Autowired val objectMapper: ObjectMapper,
+) {
 
     @MockkBean(relaxed = true)
     private lateinit var commandLineRunner: CommandLineRunner
@@ -26,19 +46,124 @@ class TagsControllerTest(@Autowired val mockMvc: MockMvc) {
     @MockkBean private lateinit var mediaFolderRepository: MediaFolderRepository
     @MockkBean private lateinit var mediaFileRepository: MediaFileRepository
 
+    private val ipAddress = InetAddress.getLocalHost().hostAddress
+
+    @Test
+    fun `Given tags, when get all endpoint called, then return them`(){
+
+        val tags = listOf(TagEntity("my_tag"))
+        every { tagsRepository.findAll() } returns tags
+
+        val result = mockMvc.perform(get("/tags"))
+            .andDo(print())
+            .andExpect(status().isOk)
+            .andReturn()
+
+        val resultResponse = objectMapper.readValue(result.response.contentAsString, object: TypeReference<List<TagEntity>>() {})
+
+        assertThat(resultResponse).hasSize(1)
+        val tagResult = resultResponse.first()
+        val expectedTag = tags.first()
+        assertTagEqual(expectedTag, tagResult)
+    }
+
     @Test
     fun `Given no tags, when add endpoint called, then create new tag`(){
 
-        every { tagsRepository.save(any()) } returns TagEntity()
+        val tag = TagEntity("my_tag")
+        every { tagsRepository.save(any()) } returns tag
 
-        mockMvc.perform(post("/tags/add?name=Cats"))
+        val result = mockMvc.perform(post("/tags/add").param("name", "Cats"))
             .andExpect(status().isOk)
-            .andExpect { content().string("Saved") }
+            .andReturn()
+
+        val resultResponse = objectMapper.readValue(result.response.contentAsString, TagEntity::class.java)
+        assertTagEqual(tag, resultResponse)
+    }
+
+    @Test
+    fun `Given valid file id, when add tag endpoint called, then add tag`(){
+
+        every { mediaFileRepository.findById(0) } returns Optional.of(ImageEntity("saved_image.jpg"))
+        val savedTag = TagEntity("my_tag")
+        every { tagsRepository.findByName("my_tag") } returns savedTag
+
+        every { tagsRepository.save(savedTag) } returns savedTag
+
+        val result = mockMvc.perform(post("/files/{id}/tags", 0)
+            .contentType(MediaType.APPLICATION_JSON)
+            .content("""{"name": "my_tag"}"""))
+            .andExpect(status().isOk)
+            .andReturn()
+
+        val resultResponse = objectMapper.readValue(result.response.contentAsString, TagEntity::class.java)
+        assertTagEqual(savedTag, resultResponse)
+    }
+
+    @Test
+    fun `Given valid tag id, when get files by tag endpoint called, then files returned`(){
+
+        every { tagsRepository.existsById(0) } returns true
+        val file = MediaFile("my_file.jpg", folder = MediaFolder("my_folder"))
+        every { mediaFileRepository.findFilesByTagsId(0, DEFAULT_PAGEABLE) } returns PageImpl(listOf(file), DEFAULT_PAGEABLE, 1)
+
+        val result = mockMvc.perform(get("/tags/{id}/files", 0))
+            .andExpect(status().isOk)
+            .andReturn()
+
+        val resultResponse = objectMapper.readValue(result.response.contentAsString, object: TypeReference<SimplePage<ImageFileDTO>>() {})
+        val expected = SimplePage(listOf(mapper.toDtoWithHost(file, ipAddress)), 1, 1)
+        assertThat(resultResponse).isEqualTo(expected)
+    }
+
+    @Test
+    fun `Given valid folder id, when get tags of folder endpoint called, then tags returned`(){
+        val folderId = 0L
+        val tag = TagEntity("my_tag")
+        val files = listOf(MediaFile("image.jpg", tags = mutableSetOf(tag)))
+        every { mediaFolderRepository.findByIdOrNull(folderId) } returns MediaFolder("my_folder", files, folderId)
+
+        val result = mockMvc.perform(get("/folders/{id}/tags", 0))
+            .andExpect(status().isOk)
+            .andReturn()
+
+        val resultResponse = objectMapper.readValue(result.response.contentAsString, object: TypeReference<List<TagEntity>>() {})
+        assertThat(resultResponse).hasSize(1)
+        val tagResult = resultResponse.first()
+        assertTagEqual(tag, tagResult)
+    }
+
+    @Test
+    fun `Given valid folder and tag ids, when get files by folder and tag endpoint called, then files returned`(){
+        val tagId = 3L
+        every { tagsRepository.existsById(tagId) } returns true
+        val folderId = 2L
+        every { mediaFolderRepository.existsById(folderId) } returns true
+        val files = listOf(MediaFile("image.jpg"))
+        every { mediaFileRepository.findFilesByTagsIdAndFolderId(tagId, folderId, DEFAULT_PAGEABLE) } returns PageImpl(files, DEFAULT_PAGEABLE, files.size.toLong())
+
+        val result = mockMvc.perform(get("/folders/{folderId}/tags/{tagId}", folderId, tagId))
+            .andExpect(status().isOk)
+            .andReturn()
+
+        val resultResponse = objectMapper.readValue(result.response.contentAsString, object: TypeReference<SimplePage<ImageFileDTO>>() {})
+        val expected = SimplePage(files.map { mapper.toDtoWithHost(it, ipAddress) }, 1, 1)
+        assertThat(resultResponse).isEqualTo(expected)
+    }
+
+    private fun assertTagEqual(expected: TagEntity, actual: TagEntity){
+        assertThat(actual.id).isEqualTo(expected.id)
+        assertThat(actual.name).isEqualTo(expected.name)
+        assertThat(actual.creationDate).isEqualTo(expected.creationDate)
     }
 
     @TestConfiguration
     internal class InnerConfig{
         @Bean
         fun fileMapper() = FileMapper()
+    }
+
+    companion object {
+        val DEFAULT_PAGEABLE = PageRequest.of(0, 20)
     }
 }

